test/e2e/util_test.go

```diff
@@ -545,20 +563,50 @@ func verifyInternalIngressController(t *testing.T, name types.NamespacedName, ho
 		"--retry-delay", "20",
 		"--max-time", "10",
 	}
-	clientPod := buildCurlPod("curl-"+name.Name, echoRoute.Namespace, image, address, echoRoute.Spec.Host, extraArgs...)
+	clientPodName := types.NamespacedName{echoRoute.Namespace, "curl-" + name.Name}
+	clientPodSpec := buildCurlPod(clientPodName.Name, clientPodName.Namespace, image, address, echoRoute.Spec.Host, extraArgs...)
+	clientPod := clientPodSpec.DeepCopy()
 	if err := kclient.Create(context.TODO(), clientPod); err != nil {
-		t.Fatalf("failed to create pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+		t.Fatalf("failed to create pod %q: %v", clientPodName, err)
 	}
 	defer func() {
 		if err := kclient.Delete(context.TODO(), clientPod); err != nil {
 			if errors.IsNotFound(err) {
 				return
 			}
-			t.Fatalf("failed to delete pod %s/%s: %v", clientPod.Namespace, clientPod.Name, err)
+			t.Fatalf("failed to delete pod %q: %v", clientPodName, err)
 		}
 	}()
 
-	err = wait.PollImmediate(10*time.Second, 5*time.Minute, func() (bool, error) {
+	curlPodLogs := ""
+	err = wait.PollImmediate(10*time.Second, 10*time.Minute, func() (bool, error) {
+		if err := kclient.Get(context.TODO(), clientPodName, clientPod); err != nil {
+			t.Logf("error getting client pod %q: %v, retrying...", clientPodName, err)
+			return false, nil
+		}
+		// First check if client curl pod is still starting or not running.
+		if clientPod.Status.Phase == corev1.PodPending {
+			t.Logf("waiting for client pod %q to start", clientPodName)
+			return false, nil
+		} else if clientPod.Status.Phase == corev1.PodFailed {
+			// If not running, the pod most likely exited, but didn't provide a 200 response, let's try again.
```

Comment:
I'm not sure we need to restart.  Do we really gain a better test with a restart, or with some log that tells us why the pod failed?

Reply:
Yes it does make the test better. There is an issue in which the loadbalancer is still initializing and will return an EOF instead of a 200 return code for a small period of time. The curl pod will exit and not try again. This will flake the test.

This is protecting against a flake, not an real-life cluster-ingress-operator issue. I provide curl pod logs regardless if the fails even after a restart.

Reply:
Why not just add a little bit of code to wait until the loadbalancer is stable?

Reply:
Oh and I should also mention, I'm planning to redo this logic on shift-week :crossed_fingers:, like @Miciah suggested.

This logic is gnarly and has edge cases I'm trying to address here, but a refactoring hopefully should make this look a lot better.

Reply:
> Why not just add a little bit of code to wait until the loadbalancer is stable before creating the pod?

We already have a `waitForIngressControllerCondition` before the call to `verifyInternalIngressController`.  What code do you have in mind?